### PR TITLE
Added TCIAInput

### PR DESCRIPTION
**Pull request by dalmouiee**
Hey guys @pchlap, @rnfinnegan, @ayhaidar 

I've added an Input class for collecting data from TCIA using their API. Currently, what this does is a Series Instance UID is provided to the Input class and used to download the DICOMs to the local filesystem. For now, this is done on a series instance basis, but it can be easily extended to grab Series UID by running a few more queries to the API by filtering Patient IDs, collections, Study IDs, etc...

Let me know what you think and how we could possibly extend this, thanks :)

### CODE REVIEW

**Comment by pchlap** [comment]
This is great @dalmouiee 

I will have a proper look later. But just wanted to let you know we also have some code in platipy to fetch from TCIA... Not sure if it makes sense to combine any of that: https://github.com/pyplati/platipy/blob/master/platipy/dicom/download/tcia.py

I'll do a proper code review over the weekend or on Monday :)

File: pydicer/input/tcia.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
+import os
+import urllib.request, urllib.error, urllib.parse
+import urllib.request, urllib.parse, urllib.error
+from pydicer.input.base import InputBase
```
Comment (pchlap) [inline pydicer/input/tcia.py:4]: Starting to nitpick here, but imports should be ordered by system python libraries first, then third party-libraries, then local imports...

File: pydicer/input/tcia.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        """
+        Input class that interfaces with the TCIA API
+
+        Args:
+            series_instance_uid (str): The series Instance UID to be downloaded from TCIA
+            working_directory (str|pathlib.Path, optional): The working directory in which to
+            store the downloaded data. Defaults to a temp directory
+        """
+        super().__init__(working_directory)
+        self.prefix_url = "https://services.cancerimagingarchive.net/services/v4/TCIA"
```
Comment (pchlap) [inline pydicer/input/tcia.py:49]: Pull this URL out as a constant?

File: pydicer/input/tcia.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+    def fetch_data(self, output_zip_file="tcia_input_zipfile.zip"):
+        """
+        Function to download the Series instance UID from TCIA and write locally
+
+        Args:
+            output_zip_file (str): name of temporary zipfile
+        """
+
+        serviceUrl = self.prefix_url + "/query/getImage"
```
Comment (pchlap) [inline pydicer/input/tcia.py:60]: I would prefer to use f-string for string concatenation

File: pydicer/input/tcia.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    query_parameters = dict((k, v) for k, v in query_parameters.items() if v)
+    queryString = "?%s" % urllib.parse.urlencode(query_parameters)
+    requestUrl = url + queryString
+    request = urllib.request.Request(url=requestUrl)
+    resp = urllib.request.urlopen(request)
+    return resp
+
+
+class TCIAInput(InputBase):
+    def __init__(self, series_instance_uid, working_directory=None):
```
Comment (pchlap) [inline pydicer/input/tcia.py:39]: Rather than specifying the series instance UID here, can we specify a TCIA collection, and a single or list of patients to fetch from the collection? An additional option could be the modalities that we are interested in.

File: pydicer/input/tcia.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        Args:
+            series_instance_uid (str): The series Instance UID to be downloaded from TCIA
+            working_directory (str|pathlib.Path, optional): The working directory in which to
+            store the downloaded data. Defaults to a temp directory
+        """
+        super().__init__(working_directory)
+        self.prefix_url = "https://services.cancerimagingarchive.net/services/v4/TCIA"
+        self.series_instance_uid = series_instance_uid
+
+    def fetch_data(self, output_zip_file="tcia_input_zipfile.zip"):
```
Comment (pchlap) [inline pydicer/input/tcia.py:52]: I don't think we need to be able to specify the output zip file here do we? Since it seems it is just used temporarily and removed after it is unzipped. Perhaps use tempfile to store it in the system temp directory?

File: tests/test_input.py [hunk 6]
```diff
@@ -12,16 +14,42 @@ def test_input_valid_working_dir_():
     # Assert path to DICOMs exists
     assert valid_filesystem_input.working_directory.is_dir()
 
+    valid_tcia_input = TCIAInput(series_instance_uid="")
+    # Assert path to DICOMs exists
+    assert os.path.isdir(valid_tcia_input.working_directory)
```
Comment (pchlap) [inline tests/test_input.py:19]: Can use pathlib builtin for this

File: tests/test_input.py [hunk 7]
```diff
@@ -12,16 +14,42 @@ def test_input_valid_working_dir_():
...
+                name
+                for name in os.listdir(invalid_series_uid_tcia_input.working_directory)
+                if os.path.isfile(name)
+            ]
+        )
+        == 0
+    )
+
 
 def test_test_input():
```
Comment (pchlap) [inline tests/test_input.py:54]: Sorry I know thats not part of this pull request, but the test_test_input function should be calling fetch_data on the test_input object, not the download_and_extract_zip_file function directly.

**Review by pchlap — CHANGES_REQUESTED** [review]
This is great stuff Daniel. I am curious how long it takes to download the images from TCIA for you? In my experience its a bit slow, but still really useful. I've got a few suggestions for changes in the comments, let me know if you want to discuss any of it in more detail.

File: pydicer/input/tcia.py (hunk 4 above)
Comment (pchlap) [reply pydicer/input/tcia.py:39]: Since in pydicer we generally are looking at grabbing all of the data for a patient, since we don't yet know exactly what we want to use from it until after we have converted to Nifti etc... :)

File: pydicer/input/tcia.py (hunk 5 above)
Comment (dalmouiee) [reply pydicer/input/tcia.py:52]: Yeah great idea!

File: pydicer/input/tcia.py (hunk 1 above)
Comment (dalmouiee) [reply pydicer/input/tcia.py:4]: Never knew about this convention, fixed for all files now!

File: pydicer/input/tcia.py (hunk 4 above)
Comment (dalmouiee) [reply pydicer/input/tcia.py:39]: Yeah for sure, I was unsure at first how well-defined a user wanted to go with the api query so I chose the lowest form as a start, but makes sense to go by collection/patient/modality(opt) as a complete form, easy change

Comment (dalmouiee) [reply pydicer/input/tcia.py:39]: Ok turns out the platipy tcia function already does this haha. I've re-written my implementation of it to include collection/patient/modality and tested both platipy's and mine to see which is faster for a range of inputs. There is any significant difference between the 2 and Platipy handles the cases we want already. So I think we should just revert to using platipy's tcia.fetch_data()

**Comment by dalmouiee** [comment]
Thanks @pchlap, I updated the TCIA one with Platipy's, both methods were relatively the same in terms of how fast they did this task. The platipy one can immediately convert to Nifty, so I imagine that we want it to keep the output in DICOM and still follow the pipeline like the other input methods so it can be preprocessed appropriately (handle missing slices, etc... )

File: tests/test_input.py [hunk 8]
```diff
@@ -12,22 +15,65 @@ def test_input_valid_working_dir_():
...
+    invalid_test_input = WebInput(data_url="", working_directory="INVALID_PATH")
     # Assert path to DICOMs does not exist
     assert not invalid_test_input.working_directory.is_dir()
 
-    invalid_filesystem_input = FilesystemInput(working_directory="NOT_VALID_PATH")
+    invalid_filesystem_input = FilesystemInput(working_directory="INVALID_PATH")
     # Assert path to DICOMs does not exist
     assert not invalid_filesystem_input.working_directory.is_dir()
 
+    invalid_work_dir_tcia_input = TCIAInput(
```
Comment (pchlap) [inline tests/test_input.py:50]: Might just need to chang the name of this variable `invalid_work_dir_tcia_input`

Comment (pchlap) [reply tests/test_input.py:50]: Hmmm wait, I think I misunderstood... Never mind this is fine

File: tests/test_input.py [hunk 9]
```diff
@@ -12,22 +15,65 @@ def test_input_valid_working_dir_():
     # Assert path to DICOMs exists
     assert valid_filesystem_input.working_directory.is_dir()
 
+    valid_tcia_input = TCIAInput(collection="", patient_ids=[], modalities=[])
```
Comment (pchlap) [inline tests/test_input.py:18]: Does this test actually download a data from a collection?

Might need to split these different valid input tests into 3 separate tests, so if one fails it will be easier to work out where it's failing. We can do that in a future PR though.

**Review by pchlap — APPROVED** [review]
Awesome! Thanks @dalmouiee. What a great feature to have this in the pydicer tool!

File: tests/test_input.py (hunk 9 above)
Comment (dalmouiee) [reply tests/test_input.py:18]: This one doesn't download any data, just confirms that a valid, temporary working directory is created when the constructor is called. We can add more tests that download actual data and make them separate for each one, but these are just sanity check
